src/Rule/RuleValidatorInterface.php

```diff
@@ -0,0 +1,22 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Yiisoft\Validator\Rule;
+
+use Yiisoft\Validator\Result;
+use Yiisoft\Validator\ValidationContext;
+use Yiisoft\Validator\ValidatorInterface;
+
+interface RuleValidatorInterface
+{
+    /**
+     * Validates the value. The method should be implemented by concrete validation rules.
+     *
+     * @param mixed $value Value to be validated.
+     * @param ValidationContext|null $context Optional validation context.
+     *
+     * @return Result
+     */
+    public function validate(mixed $value, object $rule, ValidatorInterface $validator, ?ValidationContext $context = null): Result;
```

Comment:
The big drawback here is this "object":

1. There's no enforced contract between rule and validator.
2. There's no sense in having this interface: a certain validator can only work with one type of the rule and will fail with any other type.

Other drawbacks:

1. More classes. One should implement both rule and validator instead of a single class.
2. It does not follow [information expert](https://en.wikipedia.org/wiki/GRASP_(object-oriented_design)#Information_expert) principle.